randovania/games/dread/gui/dialog/ftp_uploader.py

```diff
@@ -0,0 +1,74 @@
+import dataclasses
+import ftplib
+from ftplib import FTP
+from pathlib import Path
+from typing import Optional, Callable
+
+from randovania.lib import status_update_lib
+
+
+def delete_path(ftp: FTP, path: str, progress_update: Callable[[str], None]):
+    try:
+        file_list = list(ftp.mlsd(path))
+    except ftplib.error_perm as e:
+        if str(e) == '501 Not a directory':
+            ftp.delete(path)
+            return
+        elif str(e) in ("550 No such file or directory", "501 No such directory."):
+            return
+        else:
+            raise
+
+    for nested in file_list:
+        this_file = f"{path}/{nested[0]}"
+        if nested[1]["type"] == "dir":
+            delete_path(ftp, this_file, progress_update)
+        elif nested[1]["type"] == "file":
+            ftp.delete(this_file)
+            progress_update(this_file)
+
+    ftp.rmd(path)
+    progress_update(path)
+
+
+@dataclasses.dataclass(frozen=True)
+class FtpUploader:
+    auth: Optional[tuple[str, str]]
+    ip: str
+    port: int
+    local_path: Path
+    remote_path: str
+
+    def __call__(self, progress_update: status_update_lib.ProgressUpdateCallable):
+        all_files = list(self.local_path.rglob("*"))
+
+        with FTP() as ftp:
+            ftp.connect(self.ip, self.port)
+            if self.auth is not None:
+                ftp.login(*self.auth)
+            else:
+                ftp.login()
+
+            # Delete any existing remote path
+            delete_path(ftp, self.remote_path, lambda name: progress_update(f"Deleting existing file: {name}", 0))
```

Comment:
I'd rather this use `progress_update(str, -1)` instead, if it's not gonna be updating the progress value